readme.md
# Off

Off is a lightweight format for small configuration files.
It aims to be minimal and efficient.

# Specifications

The syntax is easy to learn, and there is three important rules I should mention:

1. The value separator is space -> 0x20 (don't mistake with = ).
2. Syntax is case sensitive.
3. Every key & value should be in a new line.

Realistic example:

```lisp
;; HTTP Server configuration file
;; Author: bindh3x

;; server
host 0.0.0.0
port 8080
banner "HTTPServer 1.0"

;; privileges
run_as_root true
user root
group wheel

;; files
document_root {/var/www/html|/home/user/httpd/|/root/httpd}
index_file {index.html|index.htm|index.php|index.phtml}

;; admin contact information
admin_email admin@example.com
admin_phone +12345678
```

Off supports four primitive types: string, array, number and boolean.
by default any value will be treated as a string

## Types

1. String

String is bunch of characters grouping together.

```lisp
;; Good
my_awesome_url http://kernel.org/
email me@example.com
pgp 0xdeadbeef

;; Bad
my_awesome_url "Hello, World"
email "foo@bar"
``` 

2. Array

Array starts with { and ends with }, elements are separated by |.

```lisp
my_array {1|2|3|4|5|6|7|Hello, World|http://bindh3x.io}
```

3. Boolean

The boolean type has only two possible values: true and false. 
Use this type for simple flags that track true/false conditions. 

```lisp
;; Good
allow_user true
allow_root false

;; Bad
allow_user True
allow_root FALSE
```

4. Number

Number is integer and can be used for various operations.

```lisp
user_id 123 
age 100
```

## Comments

Comments should starts with ;.

```lisp
name john ;; This is my awesome name

; My last name
last_name doe
```

# Implementations

## C: liboff

Very minimal implementation written by bindh3x,
the API is very simple and you can find the documentation inside the off.h file

### Features

* No dynamic memory allocation.
* Very low footprint
* No dependencies!

See ./examples/ and ./tests/ for code examples.

Example:

```c
#include <stdio.h>
#include <stdlib.h>
#include "off.h"

#define MAX_STR_LEN 256

int main(int argc, char **argv) {
  offline_t l;
  FILE *fp = NULL;
  char line[MAX_STR_LEN];
  int st = 0;
  int lines = 0;

  if ( argc < 2 ) {
    fprintf(stderr, "usage: %s [file] ...\n", argv[0]);
    exit(0);
  }

  if (( fp = fopen(argv[1], "rt")) == NULL ) {
    fprintf(stderr, "cannot open: %s\n", argv[1]);
    exit(0);
  }

  while (( fgets(line, 256, fp) ) != NULL ) { 
    off_new_line(&l);
    st = off_parse_line(line, &l);
    lines++;

    if ( st == OFF_SYNTAX_INVALID ) {
      fprintf(stderr, "invalid syntax, line %d\n", lines);
      break;
    }
    else if ( st == OFF_SYNTAX_PASS ) continue;

    printf("key start: %d, end: %d\n", l.k.start, l.k.end);
    printf("value start: %d, end: %d\n", l.v.start, l.v.end);
    printf("key  : %.*s\n", l.k.end - l.k.start, line + l.k.start);
    printf("value: %.*s\n", l.v.end - l.v.start, line + l.v.start);
  }

  return 0;
}
```

### off-utils

The off-utils.c and off-utils.h files, are not officially part of liboff.
Mainly used to save time when developing with liboff.

# Hacking

For start you can read the HACKING.txt file.
If you want to add new feature, or fix something please open a pull request,
or send a patch to os@bindh3x.io with a proper description.

# Copying

[![license](https://img.shields.io/github/license/bindh3x/off.svg?style=flat-square)]()




ChangeLog
Liboff - ChangeLog
=====================

* 27-07-2017 0.1:
 - Initial version 


off.h
/* off.h
 *
 * Copyright (C) 2017 bindh3x <os@bindh3x.io>
 *
 * This software is licensed under the terms of the GNU General Public
 * License version 3, as published by the Free Software Foundation, and
 * may be copied, distributed, and modified under those terms.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 */
#ifndef _NG_OFF
#define _NG_OFF

#define OFF_COMMENT                  0x3b
#define OFF_ARRAY_OPEN               0x7b
#define OFF_ARRAY_CLOSE              0x7d
#define OFF_ARRAY_ELEMENT_SEPARATOR  0x7c
#define OFF_SYNTAX_OK 0x03
#define OFF_SYNTAX_PASS 0x02
#define OFF_SYNTAX_INVALID 0x01

typedef enum {
  OFF_STATE_IN_KEY=1,
  OFF_STATE_IN_VALUE,
  OFF_STATE_IN_ARRAY,
} offstate_t;

typedef enum {
  OFF_TYPE_ARRAY=1,
  OFF_TYPE_STRING,
  OFF_TYPE_BOOLEAN_TRUE,
  OFF_TYPE_BOOLEAN_FALSE,
  OFF_TYPE_INTEGER,
} offtype_t;

typedef struct {
  int start; /* start index */
  int end; /* end index */
  offtype_t type; /* type of the token see above OT_* */
} offtok_t;

typedef struct {
  offtok_t k; /* key */
  offtok_t v; /* value */
} offline_t;

typedef struct {
  int el; /* number of elements in the array */
  int start; /* start position */
  int end; /* end position */
  size_t size; /* length of the array */
} offtype_array_t;

/* @brief parse one line in off syntax, the line most be valid.
 * the given `token` is going to store the information for the `line`
 *
 * @param line: line to parse
 * @param l   : offline_t, line tokens contain offtok_t for key,value.
 * @return offstate_t;
 */
int off_parse_line(char *line, offline_t *l);

/* @brief reset the offline_t for reading a new line.
 * @param l: offline_t struct
 */
void off_new_line(offline_t *l);

/* @brief guess the type of a line
 * @param line: line to guess
 * @param s   : size of line
 * @return offtype_t
 */
offtype_t off_guess_type(char *line, offtok_t *tok);

/* @brief count how many time character appears in a string.
 * @param string
 * @param character (i.e: 0x20 )
 */
int off_cchr(char *s, int c);

/* @brief parse array and return information for this array
 * @param line: line with array type to parse
 * @return    : offtype_array_t sturct, for the array
 */
offtype_array_t off_parse_array(char *line);

/* @brief return token for element in array
 * @param line: line with array type to parse
 * @param idx : index for the element 
 * @param array : offtype_array_t array from off_parse_array
 * @return offtok_t for the element
 */
offtok_t off_array_get_element(char *line, int idx, offtype_array_t *array);

/* @brief get length of a token
 * @param line: line to parse
 * @param tok: offtok_t
 * @return the length of the token string.
 */
size_t off_get_toklen(char *line, offtok_t *tok);

/* @brief compare token to a string.
 * @param line: line to parse
 * @param token: offtok_t
 * @param str  : string to compare.
 * @return 0 if equal, -1 if not.
 */
int off_tokcmp(char *line, offtok_t *tok, char *str);

/* @brief check if given string is integer
 * @param line: line to parse
 * @param tok: off token to check
 * @return 1 on success.
 */
int off_isint(char *line, offtok_t *tok);

#endif /* _NG_OFF */


off-utils.c
/* off-utils.c
 *
 * off-utils is bunch of helper functions to work with liboff
 * the functions are not officially part of off, 
 * but will help you to do a simple tasks like 
 * extracting strings from tokens etc...
 *
 * Copyright (C) 2017 bindh3x <os@bindh3x.io>
 *
 * This software is licensed under the terms of the GNU General Public
 * License version 3, as published by the Free Software Foundation, and
 * may be copied, distributed, and modified under those terms.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.-
 *
 */
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "off-utils.h"
#include "off.h"

void *offutil_malloc(size_t size) {
  void *p = NULL;

  if (( p = malloc(size)) == NULL ) {
    return NULL;
  }
  return p;
}

/* @brief: clone one by one string to malloc buffer
 * @d : dest buffer
 * @s : src buffer
 * @n : size to copy
 */
void chrclone(char *d, char *s, size_t n) {
  int i = 0;
  while ( *s ) {
    memcpy(&*d, &*s, 1);
    if ( i == n ) break;
    s++;
    d++;
    i++;
  }
  *d = '\0';
}

/* @brief extract string from offtok_t into allocated buffer
 * @paran line: line to parse in off syntax.
 * @param tok: offtok_t, token to the string.
 * @param ptr: pointer to a new NULL pointer.
 * @return the address of `ptr` with the new string.
 * on fail it will return NULL
 * XXX: call free() when done.
 */
char *offutil_extract_string(char *line, offtok_t *tok, char *ptr) {
  size_t tok_len = 0;

  tok_len = off_get_toklen(line, tok);

  if (( ptr = offutil_malloc(tok_len+1) ) != NULL ) {
    chrclone(ptr, line + tok->start, tok_len);
  }

  return ptr;
}



examples/simple.c
/* simple.c
 * ./simple ../tests/simple.off 

 * Copyright (C) 2017 bindh3x <os@bindh3x.io>
 * 
 */
#include <stdio.h>
#include <stdlib.h>
#include "../off.h"

#define MAX_STR_LEN 256

int main(int argc, char **argv) {
  offline_t l;
  FILE *fp = NULL;
  char line[MAX_STR_LEN];
  int st = 0;
  int lines = 0;

  if ( argc < 2 ) {
    fprintf(stderr, "usage: %s [file] ...\n", argv[0]);
    exit(0);
  }

  if (( fp = fopen(argv[1], "rt")) == NULL ) {
    fprintf(stderr, "cannot open: %s\n", argv[1]);
    exit(0);
  }

  while (( fgets(line, 256, fp) ) != NULL ) { 
    off_new_line(&l);
    st = off_parse_line(line, &l);
    lines++;

    if ( st == OFF_SYNTAX_INVALID ) {
      fprintf(stderr, "invalid syntax, line %d\n", lines);
      break;
    }
    else if ( st == OFF_SYNTAX_PASS ) continue;

    printf("key start: %d, end: %d\n", l.k.start, l.k.end);
    printf("value start: %d, end: %d\n", l.v.start, l.v.end);
    printf("key  : %.*s\n", l.k.end - l.k.start, line + l.k.start);
    printf("value: %.*s\n", l.v.end - l.v.start, line + l.v.start);
  }

  return 0;
}



examples/dumpoff.c
/* dumpoff.c  
 * simple example for parsing off file
 *
 * Copyright (c) 2017 bindh3x <os@bindh3x.io>
 * 21-04-2017
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../off.h"

#define MAX_STR_LEN 256

void show_type(char*, offtok_t*);

void parse_array(char *line) {
  offtype_array_t ar;
  offtok_t tok;
  int i = 0;

  ar = off_parse_array(line);
  printf("elements in the array: %d\n", ar.el);

  for ( i = 0; i < ar.el; i++ ) {
    tok = off_array_get_element(line, i, &ar);
    printf("Index: %d - start: %d, end: %d\n", i, tok.start, tok.end);
    printf("Element: %.*s\n", tok.end - tok.start, line + tok.start);
    show_type(line, &tok);
  }
  putchar(0x0a);
}

void show_type(char *line, offtok_t *tok) {
  switch(tok->type) {
      case OFF_TYPE_ARRAY:
	printf("type: ARRAY\n\n");
	break;
      case OFF_TYPE_BOOLEAN_FALSE:     
	printf("type: BOOLEAN FALSE\n\n");
	break;
      case OFF_TYPE_BOOLEAN_TRUE:
	printf("type: BOOLEAN TRUE\n\n");
	break;
      case OFF_TYPE_INTEGER:
	printf("type: INTEGER\n\n");
	break;
      case OFF_TYPE_STRING:
	printf("type: STRING\n\n");
	break;
    }
}

int main(int argc, char **argv) {
  offline_t l;
  FILE *fp = NULL;
  char line[MAX_STR_LEN];
  int st = 0;
  int lines = 0;

  if ( argc < 2 ) {
    fprintf(stderr, "usage: %s [file] ...\n", argv[0]);
    exit(0);
  }

  if (( fp = fopen(argv[1], "rt")) == NULL ) {
    fprintf(stderr, "cannot open: %s\n", argv[1]);
    exit(0);
  }

  while (( fgets(line, 256, fp) ) != NULL ) { 
    off_new_line(&l);
    st = off_parse_line(line, &l);
    lines++;

    if ( st == OFF_SYNTAX_INVALID ) {
      fprintf(stderr, "invalid syntax, line %d\n", lines);
      break;
    }
    else if ( st == OFF_SYNTAX_PASS ) continue;

    printf("key start: %d, end: %d\n", l.k.start, l.k.end);
    printf("value start: %d, end: %d\n", l.v.start, l.v.end);
    printf("key  : %.*s\n", l.k.end - l.k.start, line + l.k.start);
    printf("value: %.*s\n", l.v.end - l.v.start, line + l.v.start);
    show_type(line, &l.v);
    if ( l.v.type == OFF_TYPE_ARRAY ) {
      parse_array(line);
    }
  }

  fclose(fp);
  return 0;
}


examples/readoff.c
/* readoff.c  
 * simple example for parsing off file
 * ./readoff ../tests/array.off
 *
 * Copyright (c) 2017 bindh3x <os@bindh3x.io>
 * 23-04-2017
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "../off.h"
#include "../off-utils.h"

#define MAX_STR_LEN 256

void parse_array(char *line) {
  offtype_array_t ar;
  offtok_t tok;
  char *ptr = NULL;

  ar = off_parse_array(line);
  /* Get index 2 from the array */
  tok = off_array_get_element(line, 2, &ar);
  /* compare the element value with the URL */
  if ( off_tokcmp(line, &tok, "https://jdoe.foo") == 0 ) {
    /* if there is a match extract the full string to a new heap variable */
    ptr = offutil_extract_string(line, &tok, ptr);
    printf("Token: %s\n", ptr);
    /* free always when using offutil_extract_string */
    free(ptr);  
  }
}

int main(int argc, char **argv) {
  FILE *fp;
  char buffer[MAX_STR_LEN];
  offstate_t st;
  offline_t l;

  if ( argc < 2 ) {
    fprintf(stderr, "usage: %s [file] ...\n", argv[0]);
    exit(0);
  }

  if (( fp = fopen(argv[1], "rt") ) == NULL ) {
    fprintf(stderr, "cannot open: %s\n", argv[1]);
    exit(0);
  }

  while (( fgets(buffer, sizeof(buffer), fp) ) ) {
    off_new_line(&l);
    st = off_parse_line(buffer, &l);
  
    if ( st == OFF_SYNTAX_OK )  {
      switch(l.v.type) {
	case OFF_TYPE_ARRAY:
	  parse_array(buffer);
	  break;
	case OFF_TYPE_STRING:
	case OFF_TYPE_BOOLEAN_TRUE:
	case OFF_TYPE_BOOLEAN_FALSE:
	case OFF_TYPE_INTEGER:
	  break;
      }  
    } 
  }

  fclose(fp);
  return 0;
}


off.c
/* off.c
 *
 * C implementation for the "Ois File Format".
 *
 * Copyright (C) 2017 bindh3x <os@bindh3x.io>
 *
 * This software is licensed under the terms of the GNU General Public
 * License version 3, as published by the Free Software Foundation, and
 * may be copied, distributed, and modified under those terms.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 */
#include <string.h>
#include <ctype.h>
#include "off.h"

int off_cchr(char *s, int c) {
  int i = 0;

  while ( *s ) {
    if ( *s == c ) i++;
    s++;
  }
  return i;
}

int off_parse_line(char *line, offline_t *l) {
  offstate_t st = OFF_STATE_IN_KEY;
  int i = 0;

  /* Check if the line starts with comment or new-line */
  if ( *line == OFF_COMMENT || *line == 0x0a ) {
    return OFF_SYNTAX_PASS;
  }

  /* Remove new line character from the line */
  if ( line[strlen(line)-1] == 0x0a ) {
    line[strlen(line)-1] = 0x00;
  }

  for ( i = 0; line[i]; i++ ) {
    /* Invalid syntax */
    if ( st == OFF_STATE_IN_KEY && 
      line[i] == 0x0a &&
      i > 0 ) {
      return OFF_SYNTAX_INVALID;      
    }

    if ( st == OFF_STATE_IN_KEY ) {
      if ( line[i] == 0x20 ) {
	st = OFF_STATE_IN_VALUE;
	l->k.end = i;
	l->v.start = i+1;
	continue;
      }
    }
    else if ( st ==  OFF_STATE_IN_VALUE ) {
      /* nothing */
    }
  }
  l->v.end = i;
  l->v.type = off_guess_type(line, &l->v);

  return OFF_SYNTAX_OK;
}

void off_new_line(offline_t *l) {
  l->k.start = 0;
  l->k.end   = 0;
  l->k.type  = 0;
  l->v.start = 0;
  l->v.end   = 0;
  l->v.type  = 0;
}

offtype_t off_guess_type(char *line, offtok_t *tok) {

  /* Array */
  if ( line[strlen(line)-1] == OFF_ARRAY_CLOSE && 
      off_cchr(line, OFF_ARRAY_OPEN) == 1 && 
      off_cchr(line, OFF_ARRAY_CLOSE) == 1 ) {
    return OFF_TYPE_ARRAY;   
  }
  /* Boolean false */
  else if ( off_tokcmp(line, tok, "false") == 0 ) {
    return OFF_TYPE_BOOLEAN_FALSE;
  }
  /* Boolean true */
  else if ( off_tokcmp(line, tok, "true") == 0 ) {
    return OFF_TYPE_BOOLEAN_TRUE;
  }
  /* Integer */
  else if ( off_isint(line, tok) ) {
    return OFF_TYPE_INTEGER;
  }
  /* by default the value is a string */
  return OFF_TYPE_STRING;
}

offtype_array_t off_parse_array(char *line) {
  int i = 0;
  offtype_array_t ar;
  offstate_t st = 0;
  ar.el = 0;
  ar.size = 0;

  for ( i = 0; line[i]; i++ ) {
    if ( line[i] == OFF_ARRAY_OPEN ) {
      st = OFF_STATE_IN_ARRAY;
      ar.start = i;
      ar.el++;
      ar.size++;
      continue;        
    }
    if ( st == OFF_STATE_IN_ARRAY ) {
      if ( line[i]  == OFF_ARRAY_ELEMENT_SEPARATOR ) ar.el++;
      else if ( line[i] == OFF_ARRAY_CLOSE ) {
	ar.end = i;
	break;
      }
      ar.size++;
    }
  }
  return ar;
}

offtok_t off_array_get_element(char *line, int idx, offtype_array_t *array) {
  offtok_t tok;
  tok.end = 0;
  tok.start = array->start;
  char *lp = &line[array->start];
  int pos = array->start; /* current position */
  int index = 0;

  while ( *lp ) {
    if ( *lp == OFF_ARRAY_ELEMENT_SEPARATOR ) {
      tok.end = pos;
      if ( idx == array->el-1 ) {
	/* if we are in the last element,
	 * set tok->end to array->end */
	tok.end = array->end;
      }
      if ( index == idx ) break;
      tok.start = pos+1; /* skip OFF_ARRAY_ELEMENT_SEPARATOR */
      index++;
    }
    lp++;
    pos++;
  }
  if ( idx == 0 ) tok.start++; /* skip OFF_ARRAY_OPEN, "{element" --> "element" */
  line[array->start] = 0x3c; /* guess_type will not detect this as an array */
  tok.type = off_guess_type(line, &tok);
  line[array->start] = OFF_ARRAY_OPEN; /* now it will be detected as an array */

  return tok; 
}

size_t off_get_toklen(char *line, offtok_t *tok) {
  return (tok->end - tok->start);
}

int off_tokcmp(char *line, offtok_t *tok, char *str) {
  size_t tok_len = off_get_toklen(line, tok);

  if ( tok_len == strlen(str) &&
   strncmp(line + tok->start, str, tok_len) == 0 ) return 0;
  return -1;
}

int off_isint(char *line, offtok_t *tok) {
  char *lp = &line[tok->start];
  int i = 0;
  
  while ( *lp ) {
    if ( isdigit(*lp )) i++; 
    lp++;
  }
  if ( i == off_get_toklen(line, tok) ) return 1;
  return 0;
}
